common/common-client/src/main/java/org/ow2/proactive/authentication/iam/JWTUtils.java

```diff
@@ -42,38 +49,141 @@ private JWTUtils() {
+            jws.setCompactSerialization(jwt);
+            jws.setKey(key);
 
+            if (!jws.verifySignature()) {
+                throw new IAMException("JWT signature verification failed. Used signature key: " + jwtSignatureKey);
             }
 
-        } catch (IOException ioe) {
-            throw new IAMException("Failed to parse JWT acquired from IAM: " + token, ioe);
+            return jws.getPayload();
+
+        } catch (JoseException je) {
+            throw new IAMException("Decoding signed JWT failed", je);
         }
+    }
 
-        return claims;
+    /**
+     * Parse signed JWT (Json Web Token).
+     * @param jwt JWT (Json Web Token) to decode.
+     *
+     * @return Base-64 decoded Json Web Token.
+     */
+    private static String decodeJWT(String jwt) {
+
+        byte[] decodedBytes = Base64.decodeBase64(jwt.getBytes(StandardCharsets.UTF_8));
+        return new String(decodedBytes, StandardCharsets.UTF_8);
+
+    }
+
+    /**
+     * Parse encrypted JWT (Json Web Token).
+     *
+     * @param jwt encrypted JWT (Json Web Token) to parse.
+     * @param jwtEncryptionKey Key used to encrypt the token.
+     *
+     * @return plain text (decrypted) Json Web Token.
+     */
+    private static String decryptJWT(String jwt, String jwtEncryptionKey) {
+        JsonWebEncryption jwe = new JsonWebEncryption();
+
+        try {
+            JsonWebKey jsonWebKey = JsonWebKey.Factory.newJwk("\n" + "{\"kty\":\"oct\",\n" + " \"k\":\"" +
+                                                              jwtEncryptionKey + "\"\n" + "}");
+            jwe.setCompactSerialization(jwt);
+
+            jwe.setKey(new AesKey(jsonWebKey.getKey().getEncoded()));
+
+            return jwe.getPlaintextString();
+
+        } catch (JoseException je) {
+            throw new IAMException("Decrypting secured JWT failed. Used encryption key: " + jwtEncryptionKey, je);
+        }
+    }
+
+    /**
+     * Parses plain text JWT (Json Web Token). If successful, returns the set of claims extracted from the token body.
+     * If unsuccessful (token is invalid or not containing all required properties), simply returns empty claims.
+     *
+     * @param jwt  Plain text JWT (Json Web Token) to parse
+     * @param issuer service that issued the JWT.
+     * @param service service to which the JWT is intended for.
+     *
+     * @return the set of claims extracted from the token body, or empty claims if the token is invalid.
+     */
+    private static JwtClaims parsePlainJWT(String jwt, String issuer, String service) {
+
+        try {
+
+            JsonWebSignature jws = new JsonWebSignature();
+            jws.setAlgorithmHeaderValue(AlgorithmIdentifiers.NONE);
+            jws.setAlgorithmConstraints(AlgorithmConstraints.ALLOW_ONLY_NONE);
+            jws.setPayload(jwt);
+
+            JwtConsumer jwtConsumer = new JwtConsumerBuilder().setRequireExpirationTime() // the JWT must have an expiration time
```

Comment:
is the expiration time configurable ?

Reply:
Also, is it sensible to machine clock differences ?

Reply:
I see in https://bitbucket.org/b_c/jose4j/wiki/JWT%20Examples things like :  
jwtConsumer .setAllowedClockSkewInSeconds(30)  // allow some leeway in validating time based claims to account for clock skew

Isn't it very difficult to handle ?

Reply:
A related dicussion:
https://github.com/mozilla/addons-server/issues/1072

This looks like a serious design flaw, must be investigated

Reply:
- The expiration time is configurable at the server side (i.e., in IAM microservice). 
- As IAM makes part of PA (both are executed in the same machine), it is supposed that there is no clock differences.
- Why do you consider the clock skew is difficult to handle?

Reply:
Ok I understand, thanks